### ISSUE TITLE
The states depended on fetch trait are not correct

### ISSUE THREAD

**Issue by MrWindlike** [webzard-io/sunmao-ui#356]
**Version**
0.5.2

**Steps to reproduce**
When the fetch trait depends on other components' states behind it, it would be executed first and can't take the right state values.

The reproduce JSON schema:
```json
{
  "version": "sunmao/v1",
  "kind": "Application",
  "metadata": {
    "name": "some App"
  },
  "spec": {
    "components": [
      {
        "id": "Fetch",
        "type": "core/v1/dummy",
        "properties": {},
        "traits": [
          {
            "type": "core/v1/fetch",
            "properties": {
              "url": "/?value={{localStorage.value}}",
              "method": "get",
              "lazy": false,
              "headers": {},
              "body": {},
              "bodyType": "json",
              "onComplete": [],
              "onError": []
            }
          }
        ]
      },
      {
        "id": "localStorage",
        "type": "core/v1/dummy",
        "properties": {},
        "traits": [
          {
            "type": "core/v1/localStorage",
            "properties": {
              "key": "value",
              "value": "value"
            }
          }
        ]
      }
    ]
  }
}
```

**What is expected?**
The states' values depending on fetch trait should be correct whatever the sort is. At the above demo, the URL when fetching should be `/?value=value`.

**What is actually happening?**
![image](https://user-images.githubusercontent.com/25414733/159874706-57a5b3ba-eef5-4ba6-ae1f-4eda22b84682.png)

**Comment by tanbowensg** [comment]
The root of this bug is that during the initialization of the application, the earlier components use the state of the later components. We could solve this problem with `initStateAndMethod` to some extent. But when it comes to the situations like `defaultValue` of the `input` component, it is does not help, because the `defaultValue` of the input is not declared on the spec, but in the property, and the time to merge it on the store is during the component rendering.

So for now, one way I can think of is based on order. We can build a dependency tree of components and reorder component according it. But I haven't decide whether this sorting logic is in editor or in runtime.

However, this method cannot cover all possibilities after all, such as:

1. Expression circular dependency of two components (this may be the wrong usage)
2. The parent component uses the state of the child component

**Comment by MrWindlike** [comment]
> Could we set the defaultValue in the component spec for the states and init the states' values in initStateAndMethod to solve it?

Oh, I'm wrong. The component's `defaultValue` is a component property and its value depends on the schema.

**Comment by tanbowensg** [comment]
No, we can't fix this by reordering because it will broke the application structure. Components order matters in sunmao-ui.

**Comment by tanbowensg** [comment]
In the lastest design, fetch trait can send request  every time when component updates and we can use `disabled` property to prevent the requests which we don't want.
